### PR TITLE
[PM-9113] Trim Whitespace from email in sponsorship form

### PR DESCRIPTION
**Pull request by cturnbull-bitwarden**
## 🎟️ Tracking

https://bitwarden.atlassian.net/browse/PM-9113

## 📔 Objective

This PR aims to trim whitespace from the email field when setting up a families sponsorship. It accomplishes this by adding a validator which prevents whitespace from being added as a value.

Addtionally, this PR moves the sponsorship files to the billing folder.

## ⏰ Reminders before review

- Contributor guidelines followed
- All formatters and local linters executed and passed
- Written new unit and / or integration tests where applicable
- Protected functional changes with optionality (feature flags)
- Used internationalization (i18n) for all UI strings
- CI builds passed
- Communicated to DevOps any deployment requirements
- Updated any necessary documentation (Confluence, contributing docs) or informed the documentation team

## 🦮 Reviewer guidelines

- 👍 (`:+1:`) or similar for great changes
- 📝 (`:memo:`) or ℹ️ (`:information_source:`) for notes or general info
- ❓ (`:question:`) for questions
- 🤔 (`:thinking:`) or 💭 (`:thought_balloon:`) for more open inquiry that's not quite a confirmed issue and could potentially benefit from discussion
- 🎨 (`:art:`) for suggestions / improvements
- ❌ (`:x:`) or ⚠️ (`:warning:`) for more significant problems or concerns needing attention
- 🌱 (`:seedling:`) or ♻️ (`:recycle:`) for future improvements or indications of technical debt
- ⛏ (`:pick:`) for minor or nitpick changes

### CODE REVIEW

**Review by eliykat — COMMENTED** [review]
I don't think a validator is the right tool for this - it's not validating the form input (it never returns a ValidationError), it's mutating it. I think this is better suited to a directive, [like in this example](https://netbasal.com/how-to-trim-the-value-of-angulars-form-control-87660941e6cb). However it may also be worth talking to the CL team about in case it's functionality that can be incorporated into our CL forms.

**Comment by cturnbull-bitwarden** [comment]
> I don't think a validator is the right tool for this - it's not validating the form input (it never returns a ValidationError), it's mutating it. I think this is better suited to a directive, [like in this example](https://netbasal.com/how-to-trim-the-value-of-angulars-form-control-87660941e6cb). However it may also be worth talking to the CL team about in case it's functionality that can be incorporated directly into our CL forms.

Thanks for this, I definitely should have checked that we didn't already have a directive that did this. Turns out we did, so I pushed an update to just use that instead!

**Review by eliykat — APPROVED** [review]
Nice work!

**Review by dani-garcia — APPROVED** [review]

**Review by amorask-bitwarden — APPROVED** [review]

**Comment by cturnbull-bitwarden** [comment]
@eliykat @dani-garcia @amorask-bitwarden Hey folks, I just pushed a change addressing an issue where, if the user pressed Enter after entering an email address with spaces anywhere, the directive wouldn't update the underlying control value. I resolved this by explicitly updating both the element and the control inside the directive whenever the input event fires. Let me know your thoughts, or if you think there's a better way to do this!

Commit here: https://github.com/bitwarden/clients/pull/9781/commits/0d4bf012c31744d54b3e86db34f5f12d9d4bee45

**Review by amorask-bitwarden — DISMISSED** [review]

**Review by dani-garcia — DISMISSED** [review]

File: libs/angular/src/directives/input-strip-spaces.directive.ts [hunk 1]
```diff
@@ -1,12 +1,19 @@
...
+  constructor(
+    private el: ElementRef<HTMLInputElement>,
+    private renderer: Renderer2,
+    @Self() private ngControl: NgControl,
+  ) {}
 
   @HostListener("input") onInput() {
-    this.el.nativeElement.value = this.el.nativeElement.value.replace(/ /g, "");
+    const value = this.el.nativeElement.value.replace(/\s+/g, "");
+    this.renderer.setProperty(this.el.nativeElement, "value", value);
```
Comment (eliykat) [inline libs/angular/src/directives/input-strip-spaces.directive.ts:16]: I don't think this `renderer` call is necessary - `ngControl.control.setValue` should propagate back to the HTML element on its own. (`emitEvent: false` only prevents `valueChanges` from emitting, not the update of the native element)

File: libs/angular/src/directives/input-strip-spaces.directive.ts [hunk 2]
```diff
@@ -1,12 +1,19 @@
...
+    this.ngControl.control.setValue(value, { emitEvent: false });
```
Comment (eliykat) [inline libs/angular/src/directives/input-strip-spaces.directive.ts:17]: `emitEvent: false` means that subscribers to `valueChanges` aren't going to get the updated value after the spaces are removed. However, I see that removing it results in duplicate emissions. I think that suggests the directive approach is not the best one (my apologies), and we need to intercept the value before the FormControl updates at all - maybe in a [ControlValueAccessor](https://netbasal.com/how-to-trim-the-value-of-angulars-form-control-87660941e6cb). That could be work done by the Component Library team as part of our CL as it's a fairly common requirement and they'll know how best to implement it.

That said, I think this is non-blocking for now as most forms will only get the value on form submit, and this is making a mountain out of a very small bug. But can you please post in the CL team channel as a follow-up?

**Review by eliykat — CHANGES_REQUESTED** [review]
Only the comment re: the `renderer` call is blocking.

File: libs/angular/src/directives/input-strip-spaces.directive.ts (hunk 1 above)
Comment (cturnbull-bitwarden) [reply libs/angular/src/directives/input-strip-spaces.directive.ts]: You were correct! I pushed an update removing `renderer` as well as the `emitEvent` flag. Let me know if you've got any other thoughts

File: libs/angular/src/directives/input-strip-spaces.directive.ts (hunk 2 above)
Comment (cturnbull-bitwarden) [reply libs/angular/src/directives/input-strip-spaces.directive.ts]: I followed up with Will on this one! Thanks for your time in reviewing this one, I appreciate it!

**Review by eliykat — APPROVED** [review]
Thanks @cturnbull-bitwarden !

**Review by dani-garcia — APPROVED** [review]

**Review by amorask-bitwarden — APPROVED** [review]

**Pull request merged**